### PR TITLE
add function to set cartesian speed of the end effector for a given trajectory

### PR DESCRIPTION
**Pull request by 2scholz**
See #1556 for previous discussion.
I used the solution proposed by @kwh5484 as a basis.

To not duplicate any code I made `trajectory_processing::updateTrajectory` public and used it to adjust the velocities and accelerations of the waypoints according to the new time stamps.

### CODE REVIEW

File: moveit_core/trajectory_processing/src/iterative_time_parameterization.cpp [hunk 1]
```diff
@@ -183,8 +183,6 @@ double IterativeParabolicTimeParameterization::findT2(const double dq1, const do
   return dt2;
 }
 
-namespace
```
Comment (felixvd) [inline moveit_core/trajectory_processing/src/iterative_time_parameterization.cpp]: What happened here?

File: moveit_core/trajectory_processing/test/test_trajectory_tools.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+#include <moveit/utils/robot_model_test_utils.h>
+
+// Static variables used in all tests
+moveit::core::RobotModelConstPtr RMODEL = moveit::core::loadTestingRobotModel("pr2");
+robot_trajectory::RobotTrajectory TRAJECTORY(RMODEL, "right_arm");
+
+// Initialize one-joint, straight-line trajectory
+// Can specify init/final velocity/acceleration,
+// but not all time parameterization methods may accept it.
+int initStraightTrajectory(robot_trajectory::RobotTrajectory& trajectory, double vel_i = 0.0, double vel_f = 0.0,
```
Comment (felixvd) [inline moveit_core/trajectory_processing/test/test_trajectory_tools.cpp]: Please use longer variable names, e.g. `velocity_start`, `velocity_end`

File: moveit_core/trajectory_processing/test/test_trajectory_tools.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2017, Ken Anderson
```
Comment (felixvd) [inline moveit_core/trajectory_processing/test/test_trajectory_tools.cpp]: Update and use your name please

**Comment by felixvd** [comment]
@2scholz World MoveIt Day is coming up on June 2. Is there any chance you'd come back to this?

**Comment by v4hn** [comment]
He worked in my lab and left in the meanwhile. I would not expect him to continue this, but would encourage anyone to take over and improve this patch.

**Comment by 2scholz** [comment]
Sorry for abandoning this pull request like this. I still wanted to work on this, but never got around to do it.
I am going to join World MoveIt Day tomorrow and implement the suggested changes and additions.

**Comment by 2scholz** [comment]
I made the requested changes to the code.

I also added a planning request adapter. This adapter calls `trajectory_processing::setCartesianEndEffectorSpeed` on the given trajectory.

The adapter was also added to the template files `ompl_planning_pipeline.launch.xml` and `trajopt_planning_pipeline.launch.xml` in the `moveit_setup_assistant` package. 
The correct order for the adapters in these files is the opposite of what one would expect.
The planning adapter to set the Cartesian speed needs the result from the adapter for the time parameterization. But because both adapters require the planned trajectory, they both first call the next adapters and wait until the trajectory is planned.
That means the adapter for the time parameterization needs to be listed after the Cartesian speed adapter in order to finish first.

I added a method in `move_group_interface` that makes use of the adapter.
In order to pass the values along, I also needed to add fields in `moveit_msgs`.
See ros-planning/moveit_msgs#80 for the additional pull request.
This is also the reason why Travis CI fails to build this pull request right now.

If the speed in the field in `MotionPlanRequest` is set to 0 or a negative number, the planning request adapter ignores the request. So, if the user ignores the field, the adapter should not to anything by default.

**Comment by peasant98** [comment]
Curious -- what's the status of this?

**Comment by felixvd** [comment]
I found an old review I started but I can't find time to review this properly. You can push this forward by reviewing and testing it yourself, or even writing a test (that's a smiley and a gold star).

**Comment by Thieso** [comment]
Hi @2scholz, 

I am also interested in this function so I tried it out and implemented 2 improvements

**Compute Cartesian Path integration**
It does not work with the `computeCartesianPath()` function which in my opinion would be very helpful since that is one of the main applications. 
I went ahead and implemented this for the `computeCartesianPath()` which requires also changes to the Moveit msg `GetCartesianPath`.  I think that would be a great addition to this pull request. If you agree I can provide the necessary code and also create a PR for the msg. 

**Python interface**
Furthermore, since I am working with Python a lot, I implemented the `setMaxCartesianEndEffectorSpeed` function in the `move_group` Python interface. I can also add the code for this to this pull request. 

Otherwise I can also open new pull requests once this one is merged. Not sure what the best way to handle this is. 

What do you think?

**Comment by felixvd** [comment]
I would be interested in both of the features. Given the lack of activity on this PR, I would suggest to submit another PR with your changes based on @2scholz 's commits. A squash merge should leave both of you as authors.

**Comment by v4hn** [comment]
@2scholz left our lab quite some time ago and moved on.
If anyone has the cycles (and the use cases) to develop this further, this effort is most welcome!

**Comment by v4hn** [comment]
Closed in favor of #2674 .

**Pull request closed without merging**